src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2751,7 +2752,17 @@ ConnectionStateMachine(WorkerSession *session)
 					break;
 				}
 
+				int beforePollSocket = PQsocket(connection->pgConn);
 				PostgresPollingStatusType pollMode = PQconnectPoll(connection->pgConn);
+
+				int afterPollSocket = PQsocket(connection->pgConn);
+
+				if (beforePollSocket != afterPollSocket)
```

Comment:
Is there no possibility that even though PqconnectPoll changes the underlying connection the socket id is the same?

Reply:
I considered that and should have probably added a comment.

it does not seem like a problem, we can pass the same fd/socket to epoll_ctl. What we're interested in is not to pass any invalid sockets. And, rebuildeventset doesn't add those to the wait event set anyway. Plus, it is not possible to get invalid socket after polling.

what we do here is to eliminate old/invalid sockets/fds.

Reply:
So just to verify that I understand this correctly:

It seems that socket id is set with:
```
conn->sock = socket(addr_cur->ai_family, SOCK_STREAM, 0);
```

socket is a system call and from the documentation:

>The socket system call returns an entry into the file descriptor table (i.e. a small integer).

So we can consider the conn->sock as the file id which is used for the socket.

We would get `epoll_ctl failed no file found` error when we would try to use a connection which is not valid anymore. As in it is removed hence the file does not exit. 

So when PQconnectionPoll changes the underlying connection it means that we will have a new socket, hence a new file.

Even if the new file has the same filedescriptor id(which is probably not possible) then 
since it corresponds to an existing file, when we try to use it we wont get the `epoll_ctl failed no file found` error because there is actually a file.

But in case the file descriptor id changes, it indicates that the old file does not exit (since it is closed) so we rebuild.

Do I understand it correctly?

Reply:
yes, that's accurate. thanks for documenting the discussion here:)